core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryActivity.kt

```diff
@@ -0,0 +1,199 @@
+        mode: ActionMode,
+        menu: Menu
+      ): Boolean {
+        mode.menuInflater.inflate(R.menu.menu_context_delete, menu)
+        history_switch.isEnabled = false
+        return true
+      }
+
+      override fun onPrepareActionMode(
+        mode: ActionMode,
+        menu: Menu
+      ): Boolean {
+        return false
+      }
+
+      override fun onActionItemClicked(
+        mode: ActionMode,
+        item: MenuItem
+      ): Boolean {
+
+        if (item.itemId == id.menu_context_delete) {
+          dialogShower.show(DeleteHistory, {
+            historyViewModel.actions.offer(
+              DeleteHistoryItems(historyAdapter.items.filterIsInstance<HistoryItem>().filter { it.isSelected })
+            )
+            historyViewModel.actions.offer(Action.ExitActionModeMenu)
+            mode.finish()
+          })
+          return true
+        }
+        historyViewModel.actions.offer(Action.ExitActionModeMenu)
+        return false
+      }
+
+      override fun onDestroyActionMode(mode: ActionMode) {
+        historyViewModel.actions.offer(Action.ExitActionModeMenu)
+        history_switch.isEnabled = true
+        actionMode = null
+      }
+    }
+
+  private val historyAdapter: HistoryAdapter2 by lazy {
+    HistoryAdapter2(
+      HistoryItemDelegate(this)
+    )
+  }
+
+  override fun injection(coreComponent: CoreComponent) {
+    activityComponent.inject(this)
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_history)
+    setSupportActionBar(toolbar)
+
+    supportActionBar?.setDisplayHomeAsUpEnabled(true)
+    supportActionBar?.setTitle(string.history)
+
+    recycler_view.adapter = historyAdapter
+    recycler_view.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
+
+    history_switch.setOnCheckedChangeListener { _, isChecked ->
+      historyViewModel.actions.offer(ToggleShowHistoryFromAllBooks(isChecked))
+    }
+    history_switch.isChecked = !sharedPreferenceUtil.showHistoryCurrentBook
+
+    compositeDisposable.add(historyViewModel.effects.subscribe { it.invokeWith(this) })
+  }
+
+  override fun onDestroy() {
+    compositeDisposable.clear()
+    super.onDestroy()
+  }
+
+  override fun onCreateOptionsMenu(menu: Menu): Boolean {
+    menuInflater.inflate(R.menu.menu_history, menu)
+    val search = menu.findItem(id.menu_history_search).actionView as SearchView
+    search.queryHint = getString(string.search_history)
+    search.setOnQueryTextListener(SimpleTextListener {
+      historyViewModel.actions.offer(Filter(it))
+    })
+    historyViewModel.state.observe(this, Observer(::render))
+    return true
+  }
+
+  private fun render(state: State) =
+    when (state) {
+      is Results -> {
+        if (state.historyItems.filter { it is HistoryItem && it.isSelected }
+            .isNotEmpty() && actionMode == null) {
+          actionMode = startSupportActionMode(actionModeCallback)
```

Comment:
The view is starting to look very nice and simple, some stuff to be deleted of course and all the things happening in the action mode callback turned into SideEffects.

This is the part that stumped me, actionModes. I never really spent enough time to see if there was a re-entrant way of invoking it. If there was a way of retrieving action modes it would be a lot better but no you have to store the variable returned from this damn call. Anyway I can allow this even if it isn't technically pure MVI.

Also of note is that I have an extension function for starting action modes, see core/ActivityExtensions

Reply:
Indeed it is, I finally felt like I got a grasp of how nice this pattern is/how to use rxjava with the viewmodel. It's a lot of going back and fourth between the activity, viewmodel, actions, states, ...... 🦖 

I'll take a look at the extension function and fix side effects. Thanks for your help & feedback again!

Reply:
Unidirectional Data Flow, when correct it is a perfect circle. We don't take it particularly far in kiwix but there are plenty of time rewinding debuggers out there, [MVIKotlin](https://github.com/arkivanov/MVIKotlin) just went multiplatform, very cool

Reply:
Time rewinding debuggers?? That is cool! 

For the startSupportActionMode extension: it exits before the dialog has a chance to take effect. I.e. all items are deselected and the action mode disappears as soon as the dialog is opened. 

```
private fun requestDeletionOfSelectedItems(){
    dialogShower.show(DeleteHistory, {
      historyViewModel.actions.offer(
        DeleteHistoryItems(historyAdapter.items.filterIsInstance<HistoryItem>().filter { it.isSelected })
      )
      historyViewModel.actions.offer(Action.ExitActionModeMenu)
    })
  }
...
startActionMode(
  R.menu.menu_context_delete,
  mapOf(id.menu_context_delete to { requestDeletionOfSelectedItems() }),
  ::exitSelectionActionMode
)
```
in ActivityExtensions.startActionMode: 

```
override fun onActionItemClicked(mode: ActionMode, item: MenuItem) =
        idsToClickActions[item.itemId]?.let {
          it()
          mode.finish()
          true
        } ?: false

```

One solution is to let the users of startActionMode decide when to finish the actionMode or do you have any other suggestions? I could also skip using the extension.

Reply:
hmmm I guess you can avoid the extension method, it could probably be improved but I don't want to go fixing ZimFileSelectFragment